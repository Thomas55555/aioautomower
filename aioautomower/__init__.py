"""Automower library using aiohttp."""
from .rest import *
from .session import AutomowerSession
