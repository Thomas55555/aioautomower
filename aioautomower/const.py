"""The constants for aioautomower."""

API_BASE_URL = "https://api.amc.husqvarna.dev/v1"
AUTH_API_BASE_URL = "https://api.authentication.husqvarnagroup.dev/v1"
AUTH_API_TOKEN_URL = f"{AUTH_API_BASE_URL}/oauth2/token"
AUTH_API_REVOKE_URL = f"{AUTH_API_BASE_URL}/oauth2/revoke"
AUTH_HEADERS = {
    "Content-Type": "application/x-www-form-urlencoded",
    "Accept": "application/json",
}
AUTH_HEADER_FMT = "Bearer {}"
EVENT_TYPES = [
    "status-event",
    "positions-event",
    "settings-event",
]
HUSQVARNA_URL = "https://developer.husqvarnagroup.cloud/"
MARGIN_TIME = 60.0  # Token is typically valid for 24h, request a new one some time before its expiration to avoid glitches.
MIN_SLEEP_TIME = 10.0  # Avoid hammering
REST_POLL_CYCLE = 300
TOKEN_URL = f"{AUTH_API_BASE_URL}/token"
USER_URL = f"{AUTH_API_BASE_URL}/users"
WS_URL = "wss://ws.openapi.husqvarna.dev/v1"


ERRORCODES = {
    0: "Unexpected error",
    1: "Outside working area",
    2: "No loop signal",
    3: "Wrong loop signal",
    4: "Loop sensor problem, front",
    5: "Loop sensor problem, rear",
    6: "Loop sensor problem, left",
    7: "Loop sensor problem, right",
    8: "Wrong PIN code",
    9: "Trapped",
    10: "Upside down",
    11: "Low battery",
    12: "Empty battery",
    13: "No drive",
    14: "Mower lifted",
    15: "Lifted",
    16: "Stuck in charging station",
    17: "Charging station blocked",
    18: "Collision sensor problem, rear",
    19: "Collision sensor problem, front",
    20: "Wheel motor blocked, right",
    21: "Wheel motor blocked, left",
    22: "Wheel drive problem, right",
    23: "Wheel drive problem, left",
    24: "Cutting system blocked",
    25: "Cutting system blocked",
    26: "Invalid sub-device combination",
    27: "Settings restored",
    28: "Memory circuit problem",
    29: "Slope too steep",
    30: "Charging system problem",
    31: "STOP button problem",
    32: "Tilt sensor problem",
    33: "Mower tilted",
    34: "Cutting stopped - slope too steep",
    35: "Wheel motor overloaded, right",
    36: "Wheel motor overloaded, left",
    37: "Charging current too high",
    38: "Electronic problem",
    39: "Cutting motor problem",
    40: "Limited cutting height range",
    41: "Unexpected cutting height adj",
    42: "Limited cutting height range",
    43: "Cutting height problem, drive",
    44: "Cutting height problem, curr",
    45: "Cutting height problem, dir",
    46: "Cutting height blocked",
    47: "Cutting height problem",
    48: "No response from charger",
    49: "Ultrasonic problem",
    50: "Guide 1 not found",
    51: "Guide 2 not found",
    52: "Guide 3 not found",
    53: "GPS navigation problem",
    54: "Weak GPS signal",
    55: "Difficult finding home",
    56: "Guide calibration accomplished",
    57: "Guide calibration failed",
    58: "Temporary battery problem",
    59: "Temporary battery problem",
    60: "Temporary battery problem",
    61: "Temporary battery problem",
    62: "Battery restriction due to ambient temperature",
    63: "Temporary battery problem",
    64: "Temporary battery problem",
    65: "Temporary battery problem",
    66: "Battery problem",
    67: "Battery problem",
    68: "Temporary battery problem",
    69: "Alarm! Mower switched off",
    70: "Alarm! Mower stopped",
    71: "Alarm! Mower lifted",
    72: "Alarm! Mower tilted",
    73: "Alarm! Mower in motion",
    74: "Alarm! Outside geofence",
    75: "Connection changed",
    76: "Connection NOT changed",
    77: "Com board not available",
    78: "Slipped - Mower has Slipped.Situation not solved with moving pattern",
    79: "Invalid battery combination - Invalid combination of different battery types.",
    80: "Cutting system imbalance    Warning",
    81: "Safety function faulty",
    82: "Wheel motor blocked, rear right",
    83: "Wheel motor blocked, rear left",
    84: "Wheel drive problem, rear right",
    85: "Wheel drive problem, rear left",
    86: "Wheel motor overloaded, rear right",
    87: "Wheel motor overloaded, rear left",
    88: "Angular sensor problem",
    89: "Invalid system configuration",
    90: "No power in charging station",
    91: "Switch cord problem",
    92: "Work area not valid",
    93: "No accurate position from satellites",
    94: "Reference station communication problem",
    95: "Folding sensor activated",
    96: "Right brush motor overloaded",
    97: "Left brush motor overloaded",
    98: "Ultrasonic Sensor 1 defect",
    99: "Ultrasonic Sensor 2 defect",
    100: "Ultrasonic Sensor 3 defect",
    101: "Ultrasonic Sensor 4 defect",
    102: "Cutting drive motor 1 defect",
    103: "Cutting drive motor 2 defect",
    104: "Cutting drive motor 3 defect",
    105: "Lift Sensor defect",
    106: "Collision sensor defect",
    107: "Docking sensor defect",
    108: "Folding cutting deck sensor defect",
    109: "Loop sensor defect",
    110: "Collision sensor error",
    111: "No confirmed position",
    112: "Cutting system major imbalance",
    113: "Complex working area",
    114: "Too high discharge current",
    115: "Too high internal current",
    116: "High charging power loss",
    117: "High internal power loss",
    118: "Charging system problem",
    119: "Zone generator problem",
    120: "Internal voltage error",
    121: "High internal temperature",
    122: "CAN error",
    123: "Destination not reachable",
    701: "Connectivity problem",
    702: "Connectivity settings restored",
    703: "Connectivity problem",
    704: "Connectivity problem",
    705: "Connectivity problem",
    706: "Poor signal quality",
    707: "SIM card requires PIN",
    708: "SIM card locked",
    709: "SIM card not found",
    710: "SIM card locked",
    711: "SIM card locked",
    712: "SIM card locked",
    713: "Geofence problem",
    714: "Geofence problem",
    715: "Connectivity problem",
    716: "Connectivity problem",
    717: "SMS could not be sent",
    724: "Communication circuit board SW must be updated",
}
