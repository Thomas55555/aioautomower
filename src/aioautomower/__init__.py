"""Automower library using aiohttp."""

__all__ = [
    "auth",
    "const",
    "exceptions",
    "model",
    "session",
    "utils",
]
