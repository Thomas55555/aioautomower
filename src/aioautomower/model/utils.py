"""Helper utils for Automower Connect API."""

from datetime import UTC, datetime

from aioautomower import tz_util


def convert_timestamp_to_aware_datetime(timestamp: int) -> datetime | None:
    """Convert the timestamp to an aware datetime object.

    The Python datetime library expects timestamps to be anchored in UTC,
    however, the automower timestamps are anchored in local time. So we convert
    the timestamp to a datetime and replace the timezone with the local time.
    After that we convert the timezone to UTC.
    """
    if timestamp == 0:
        return None
    if timestamp > 32503680000:
        # This will break on January 1th 3000. If mankind still exists there
        # please fix it.
        return datetime.fromtimestamp(timestamp / 1000, tz=UTC).replace(
            tzinfo=tz_util.MOWER_TIME_ZONE
        )
    return datetime.fromtimestamp(timestamp, tz=UTC).replace(
        tzinfo=tz_util.MOWER_TIME_ZONE
    )
