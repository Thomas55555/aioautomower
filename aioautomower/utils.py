"""Utils for Husqvarna Automower."""
import logging
import time
from urllib.parse import quote_plus, urlencode

import aiohttp
import jwt

from .const import AUTH_API_REVOKE_URL, AUTH_API_TOKEN_URL, AUTH_HEADERS
from .model import JWT
from .rest import TokenError

_LOGGER = logging.getLogger(__name__)


async def async_structure_token(access_token) -> JWT:
    """Decode JWT and convert to dataclass."""
    token_decoded = jwt.decode(access_token, options={"verify_signature": False})
    return JWT(**token_decoded)


async def async_get_access_token(client_id, client_secret) -> dict:
    """Get an access token from the Authentication API with client credentials.

    This grant type is intended only for you. If you want other
    users to use your application, then they should login using Authorization
    Code Grant.
    """
    auth_data = urlencode(
        {
            "grant_type": "client_credentials",
            "client_id": client_id,
            "client_secret": client_secret,
        },
        quote_via=quote_plus,
    )
    async with aiohttp.ClientSession(headers=AUTH_HEADERS) as session, session.post(
        AUTH_API_TOKEN_URL, data=auth_data
    ) as resp:
        result = await resp.json(encoding="UTF-8")
        _LOGGER.debug("Resp.status get access token: %s", result)
        if resp.status == 200:
            result = await resp.json(encoding="UTF-8")
            result["expires_at"] = result["expires_in"] + time.time()
        if resp.status >= 400:
            raise TokenError(
                f"""The token is invalid, respone from
                    Husqvarna Automower API: {result}"""
            )
    result["status"] = resp.status
    return result


async def async_invalidate_access_token(
    valid_access_token, access_token_to_invalidate
) -> dict:
    """Invalidate the token.

    :param str valid_access_token: A working access token to authorize this request.
    :param str access_token_to_delete: An access token to invalidate,
    can be th same like the first argument.
    """
    headers = {
        "Content-Type": "application/x-www-form-urlencoded",
        "Authorization": f"Bearer {valid_access_token}",
        "Accept": "*/*",
    }
    async with aiohttp.ClientSession(headers=headers) as session, session.post(
        AUTH_API_REVOKE_URL, data=(f"token={access_token_to_invalidate}")
    ) as resp:
        result = await resp.json(encoding="UTF-8")
        _LOGGER.debug("Resp.status delete token: %s", resp.status)
        if resp.status >= 400:
            resp.raise_for_status()
            _LOGGER.error("Response body delete token: %s", result)
    return result
