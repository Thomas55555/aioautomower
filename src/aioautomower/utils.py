"""Utils for Husqvarna Automower."""

import logging
import time
from typing import Any, Mapping, cast
from urllib.parse import quote_plus, urlencode

import aiohttp
import jwt

from .const import AUTH_API_REVOKE_URL, AUTH_API_TOKEN_URL, AUTH_HEADERS, ERRORCODES
from .exceptions import ApiException
from .model import JWT, MowerAttributes, MowerList, snake_case

_LOGGER = logging.getLogger(__name__)


def structure_token(access_token: str) -> JWT:
    """Decode JWT and convert to dataclass."""
    token_decoded = jwt.decode(access_token, options={"verify_signature": False})
    return JWT.from_dict(token_decoded)


async def async_get_access_token(client_id: str, client_secret: str) -> dict[str, str]:
    """Get an access token from the Authentication API with client credentials.

    This grant type is intended only for you. If you want other
    users to use your application, then they should login using Authorization
    Code Grant.
    """
    auth_data = urlencode(
        {
            "grant_type": "client_credentials",
            "client_id": client_id,
            "client_secret": client_secret,
        },
        quote_via=quote_plus,
    )
    async with (
        aiohttp.ClientSession(headers=AUTH_HEADERS) as session,
        session.post(AUTH_API_TOKEN_URL, data=auth_data) as resp,
    ):
        result = await resp.json(encoding="UTF-8")
        _LOGGER.debug("Resp.status get access token: %s", result)
        if resp.status == 200:
            result = await resp.json(encoding="UTF-8")
            result["expires_at"] = result["expires_in"] + time.time()
        if resp.status >= 400:
            raise ApiException(
                f"""The token is invalid, response from
                    Husqvarna Automower API: {result}"""
            )
    result["status"] = resp.status
    return cast(dict[str, str], result)


async def async_invalidate_access_token(
    valid_access_token: str, access_token_to_invalidate: str
) -> dict[str, str]:
    """Invalidate the token.

    :param str valid_access_token: A working access token to authorize this request.
    :param str access_token_to_delete: An access token to invalidate,
    can be th same like the first argument.
    """
    headers = {
        "Content-Type": "application/x-www-form-urlencoded",
        "Authorization": f"Bearer {valid_access_token}",
        "Accept": "*/*",
    }
    async with (
        aiohttp.ClientSession(headers=headers) as session,
        session.post(
            AUTH_API_REVOKE_URL, data=(f"token={access_token_to_invalidate}")
        ) as resp,
    ):
        result = await resp.json(encoding="UTF-8")
        _LOGGER.debug("Resp.status delete token: %s", resp.status)
        if resp.status >= 400:
            resp.raise_for_status()
            _LOGGER.error("Response body delete token: %s", result)
    return cast(dict[str, str], result)


def mower_list_to_dictionary_dataclass(
    mower_list: Mapping[Any, Any],
) -> dict[str, MowerAttributes]:
    """Convert mower data to a dictionary DataClass."""
    mowers_list = MowerList.from_dict(mower_list)
    mowers_dict = {}
    for mower in mowers_list.data:
        mowers_dict[mower.id] = mower.attributes
    return mowers_dict


def error_key_list() -> list[str]:
    """Create a list with all possible error keys."""
    codes = [snake_case(error_text) for error_text in ERRORCODES.values()]
    return sorted(codes)


def error_key_dict() -> dict[str, str]:
    """Create a dictionary with error keys and a human friendly text."""
    codes = {}
    for error_text in ERRORCODES.values():
        codes[snake_case(error_text)] = error_text
    return codes
