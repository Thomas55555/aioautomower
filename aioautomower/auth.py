"""Module for AbstractAuth for Husqvarna Automower."""

from abc import ABC, abstractmethod
from collections.abc import Mapping
from http import HTTPStatus
import logging
from typing import Any, Optional

import aiohttp
from aiohttp import ClientError, ClientResponse, ClientResponseError, ClientSession

from .const import API_BASE_URL, AUTH_HEADER_FMT, EVENT_TYPES, WS_URL
from .exceptions import ApiException, ApiForbiddenException, AuthException
from .utils import async_structure_token

ERROR = "error"
STATUS = "status"
MESSAGE = "message"

_LOGGER = logging.getLogger(__name__)


class AbstractAuth(ABC):
    """Abstract class to make authenticated requests."""

    def __init__(self, websession: ClientSession, host: str) -> None:
        """Initialize the auth."""
        self._websession = websession
        self._host = host if host is not None else API_BASE_URL
        self._client_id = None
        self.loop = websession.loop
        self.ws_update_cbs = []
        self.ws_data = {}

    @abstractmethod
    async def async_get_access_token(self) -> str:
        """Return a valid access token."""

    async def request(
        self, method: str, url: str, **kwargs: Optional[Mapping[str, Any]]
    ) -> ClientResponse:
        """Make a request."""
        headers = await self.headers()
        if not (url.startswith("http://") or url.startswith("https://")):
            url = f"{self._host}/{url}"
        _LOGGER.debug("request[%s]=%s %s", method, url, kwargs.get("params"))
        if method != "get" and "json" in kwargs:
            _LOGGER.debug("request[post json]=%s", kwargs["json"])
        return await self._websession.request(method, url, **kwargs, headers=headers)

    async def get(self, url: str, **kwargs: Mapping[str, Any]) -> ClientResponse:
        """Make a get request."""
        try:
            resp = await self.request("get", url, **kwargs)
        except ClientError as err:
            raise ApiException(f"Error connecting to API: {err}") from err
        return await AbstractAuth._raise_for_status(resp)

    async def get_json(self, url: str, **kwargs: Mapping[str, Any]) -> dict[str, Any]:
        """Make a get request and return json response."""
        resp = await self.get(url, **kwargs)
        try:
            result = await resp.json(encoding="UTF-8")
        except ClientError as err:
            raise ApiException("Server returned malformed response") from err
        if not isinstance(result, dict):
            raise ApiException(f"Server return malformed response: {result}")
        _LOGGER.debug("response=%s", result)
        return result

    async def post(self, url: str, **kwargs: Mapping[str, Any]) -> ClientResponse:
        """Make a post request."""
        try:
            resp = await self.request("post", url, **kwargs)
        except ClientError as err:
            raise ApiException(f"Error connecting to API: {err}") from err
        return await AbstractAuth._raise_for_status(resp)

    async def post_json(self, url: str, **kwargs: Mapping[str, Any]) -> dict[str, Any]:
        """Make a post request and return a json response."""
        resp = await self.post(url, **kwargs)
        try:
            result = await resp.json()
        except ClientError as err:
            raise ApiException("Server returned malformed response") from err
        if not isinstance(result, dict):
            raise ApiException(f"Server returned malformed response: {result}")
        _LOGGER.debug("response=%s", result)
        return result

    async def patch(self, url: str, **kwargs: Mapping[str, Any]) -> ClientResponse:
        """Make a post request."""
        try:
            resp = await self.request("patch", url, **kwargs)
        except ClientError as err:
            raise ApiException(f"Error connecting to API: {err}") from err
        return await AbstractAuth._raise_for_status(resp)

    async def patch_json(self, url: str, **kwargs: Mapping[str, Any]) -> dict[str, Any]:
        """Make a post request and return a json response."""
        resp = await self.patch(url, **kwargs)
        try:
            result = await resp.json()
        except ClientError as err:
            raise ApiException("Server returned malformed response") from err
        if not isinstance(result, dict):
            raise ApiException(f"Server returned malformed response: {result}")
        _LOGGER.debug("response=%s", result)
        return result

    async def _async_get_access_token(self) -> None:
        """Requests a new access token."""
        try:
            return await self.async_get_access_token()
        except ClientError as err:
            raise AuthException(f"Access token failure: {err}") from err

    async def headers(self) -> None:
        """Generate headers for ReST requests."""
        access_token = await self._async_get_access_token()
        if not self._client_id:
            token_structured = await async_structure_token(access_token)
            self._client_id = token_structured.client_id
        return {
            "Authorization": f"Bearer {access_token}",
            "Authorization-Provider": "husqvarna",
            "Content-Type": "application/vnd.api+json",
            "X-Api-Key": self._client_id,
        }

    @staticmethod
    async def _raise_for_status(resp: ClientResponse) -> ClientResponse:
        """Raise exceptions on failure methods."""
        detail = await AbstractAuth._error_detail(resp)
        try:
            resp.raise_for_status()
        except ClientResponseError as err:
            if err.status == HTTPStatus.FORBIDDEN:
                raise ApiForbiddenException(
                    f"Forbidden response from API: {err}"
                ) from err
            if err.status == HTTPStatus.UNAUTHORIZED:
                raise AuthException(f"Unable to authenticate with API: {err}") from err
            detail.append(err.message)
            raise ApiException(": ".join(detail)) from err
        except ClientError as err:
            raise ApiException(f"Error from API: {err}") from err
        return resp

    @staticmethod
    async def _error_detail(resp: ClientResponse) -> list[str]:
        """Return an error message string from the API response."""
        if resp.status < 400:
            return []
        try:
            result = await resp.json()
            error = result.get(ERROR, {})
        except ClientError:
            return []
        message = ["Error from API", f"{resp.status}"]
        if STATUS in error:
            message.append(f"{error[STATUS]}")
        if MESSAGE in error:
            message.append(error[MESSAGE])
        return message

    def register_ws_callback(self, callback):
        """Register a websocket update callback.

        :param func callback: Callback fired on data updates. Takes one dict argument which is the latest websocket data.
        """
        if callback not in self.ws_update_cbs:
            self.ws_update_cbs.append(callback)

    def _schedule_ws_callback(self, cb, delay=0.0):
        """Schedule websocket callback."""
        self.loop.call_later(delay, cb, self.ws_data)

    def _schedule_ws_callbacks(self):
        """Schedule all websocket callbacks."""
        for cb in self.ws_update_cbs:
            self._schedule_ws_callback(cb)

    async def websocket(self) -> dict[str, dict]:
        """Start a websocket conenction."""
        async with self._websession.ws_connect(
            url=WS_URL,
            headers={
                "Authorization": AUTH_HEADER_FMT.format(
                    await self._async_get_access_token()
                )
            },
            heartbeat=60,
        ) as ws:
            async for msg in ws:
                if msg.type == aiohttp.WSMsgType.TEXT:
                    msg_dict = msg.json()
                    if "type" in msg_dict:
                        if msg_dict["type"] in EVENT_TYPES:
                            _LOGGER.debug(
                                "Got %s, data: %s", msg_dict["type"], msg_dict
                            )
                            self.ws_data = msg_dict
                            self._schedule_ws_callbacks()
                        else:
                            _LOGGER.warning(
                                "Received unknown ws type %s", msg_dict["type"]
                            )
                    elif "ready" in msg_dict and "connectionId" in msg_dict:
                        _LOGGER.debug(
                            "Websocket ready=%s (id='%s')",
                            msg_dict["ready"],
                            msg_dict["connectionId"],
                        )
                    else:
                        _LOGGER.debug("Discarded websocket response: %s", msg_dict)
                elif msg.type == aiohttp.WSMsgType.ERROR:
                    _LOGGER.debug("Received ERROR")
                    break
                elif msg.type == aiohttp.WSMsgType.CONTINUATION:
                    _LOGGER.debug("Received CONTINUATION")
                elif msg.type == aiohttp.WSMsgType.BINARY:
                    _LOGGER.debug("Received BINARY")
                elif msg.type == aiohttp.WSMsgType.PING:
                    _LOGGER.debug("Received PING")
                elif msg.type == aiohttp.WSMsgType.PONG:
                    _LOGGER.debug("Received PONG")
                elif msg.type == aiohttp.WSMsgType.CLOSE:
                    _LOGGER.debug("Received CLOSE")
                    await ws.close()
                elif msg.type == aiohttp.WSMsgType.CLOSING:
                    _LOGGER.debug("Received CLOSING")
                elif msg.type == aiohttp.WSMsgType.CLOSED:
                    _LOGGER.debug("Received CLOSED")
                else:
                    _LOGGER.debug("Received msg.type=%d", msg.type)
