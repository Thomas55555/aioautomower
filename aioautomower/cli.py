#!/usr/bin/env python3
import argparse
import asyncio
import logging
import signal
import time

from aiohttp import ClientSession

from aioautomower.auth import AbstractAuth
from aioautomower.session import AutomowerSession
import aioautomower.utils

from .const import API_BASE_URL

_LOGGER = logging.getLogger(__name__)


async def run_tester(client_id: str, client_secret: str):
    automower_api = AutomowerSession(
        AsyncConfigEntryAuth(ClientSession(), client_id, client_secret), poll=True
    )
    automower_api.register_data_callback(
        lambda x: logging.info("data callback;%s", x),
    )

    await automower_api.connect()

    def sigusr1():
        asyncio.ensure_future(automower_api.invalidate_token())

    def sigusr2():
        asyncio.ensure_future(automower_api.get_status())

    loop = asyncio.get_event_loop()
    loop.add_signal_handler(signal.SIGUSR1, sigusr1)
    loop.add_signal_handler(signal.SIGUSR2, sigusr2)

    while True:
        await asyncio.sleep(0.1)


def main():
    """Tester for the Husqvarna Automower API.

    The tester will login using client_id and client_secret and connect to a
    websocket listening for mower updates.

    The tester listens to two signals on which it performs the following
    requests:
    SIGUSR1: Invalidate token
    SIGUSR2: Get status
    """

    parser = argparse.ArgumentParser(
        description=main.__doc__, formatter_class=argparse.RawTextHelpFormatter
    )
    parser.add_argument(
        "-s", "--client_id", required=True, help="Husqvarna Application key"
    )
    parser.add_argument(
        "-k", "--client_secret", required=True, help="Husqvarna Application secret"
    )

    args = parser.parse_args()

    logging.basicConfig(level="DEBUG", format="%(asctime)s;%(levelname)s;%(message)s")
    asyncio.run(run_tester(args.client_id, args.client_secret))


class AsyncConfigEntryAuth(AbstractAuth):
    """Provide Husqvarna Automower authentication tied to an OAuth2 based config entry."""

    def __init__(
        self,
        websession: ClientSession,
        client_id: str,
        client_secret: str,
    ) -> None:
        """Initialize Husqvarna Automower auth."""
        super().__init__(websession, API_BASE_URL)
        self.client_id = client_id
        self.client_secret = client_secret
        self.token = None
        self.websession = websession

    async def async_get_access_token(self) -> str:
        """Return a valid access token."""
        if not self.token:
            self.token = await aioautomower.utils.async_get_access_token(
                self.client_id, self.client_secret, self.websession
            )
            token_structured = await aioautomower.utils.async_structure_token(
                self.token["access_token"]
            )
            _LOGGER.debug("token_structured.exp: %s", token_structured.exp)
            _LOGGER.debug(time.time())
        if token_structured.exp < time.time():
            self.token = await aioautomower.utils.async_get_access_token(
                self.client_id, self.client_secret, self.websession
            )
        return self.token["access_token"]
