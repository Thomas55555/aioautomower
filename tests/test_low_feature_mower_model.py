"""Tests for asynchronous Python client for aioautomower."""

import json
from dataclasses import fields

import pytest

from aioautomower.utils import mower_list_to_dictionary_dataclass
from tests import load_fixture

MOWER_ID = "c7233734-b219-4287-a173-08e3643f89f0"


@pytest.mark.asyncio
async def test_low_feature_mower() -> None:
    """Test converting a low feature mower."""
    mower_fixture = load_fixture("low_feature_mower.json")
    mower_python = json.loads(mower_fixture)
    mowers = mower_list_to_dictionary_dataclass(mower_python)
    assert mowers[MOWER_ID].headlight.mode is None
    assert mowers[MOWER_ID].cutting_height is None
    assert len(mowers[MOWER_ID].positions) == 0


def test_mower_snapshot(snapshot):
    """Testing a snapshot of a high feature mower."""
    mower_fixture = load_fixture("low_feature_mower.json")
    mower_python = json.loads(mower_fixture)
    mowers = mower_list_to_dictionary_dataclass(mower_python)
    for field in fields(mowers[MOWER_ID]):
        field_name = field.name
        field_value = getattr(mowers[MOWER_ID], field_name)
        assert field_value == snapshot(name=f"{field_name}")
