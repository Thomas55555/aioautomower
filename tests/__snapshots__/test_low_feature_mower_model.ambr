# serializer version: 1
# name: test_mower_snapshot[battery]
  dict({
    'battery_percent': 50,
  })
# ---
# name: test_mower_snapshot[calendar]
  dict({
    'events': list([
      dict({
        'end': FakeDatetime(2024, 5, 13, 0, 49, tzinfo=datetime.timezone.utc),
        'rrule': 'FREQ=WEEKLY;BYDAY=MO',
        'start': FakeDatetime(2024, 5, 13, 0, 0, tzinfo=datetime.timezone.utc),
        'uid': '120_49_MO',
        'work_area_id': None,
      }),
    ]),
    'tasks': list([
      dict({
        'duration': 49,
        'friday': False,
        'monday': True,
        'saturday': False,
        'start': 120,
        'sunday': False,
        'thursday': False,
        'tuesday': False,
        'wednesday': False,
        'work_area_id': None,
      }),
    ]),
  })
# ---
# name: test_mower_snapshot[capabilities]
  dict({
    'headlights': False,
    'position': False,
    'stay_out_zones': False,
    'work_areas': False,
  })
# ---
# name: test_mower_snapshot[metadata]
  dict({
    'connected': True,
    'status_dateteime': FakeDatetime(2023, 10, 18, 22, 58, 52, 683000, tzinfo=datetime.timezone.utc),
  })
# ---
# name: test_mower_snapshot[mower]
  dict({
    'activity': 'PARKED_IN_CS',
    'error_code': 0,
    'error_datetime': None,
    'error_datetime_naive': None,
    'error_key': None,
    'inactive_reason': 'NONE',
    'is_error_confirmable': False,
    'mode': 'MAIN_AREA',
    'state': 'RESTRICTED',
    'work_area_id': None,
  })
# ---
# name: test_mower_snapshot[planner]
  dict({
    'next_start_datetime': FakeDatetime(2023, 6, 5, 17, 0, tzinfo=datetime.timezone.utc),
    'next_start_datetime_naive': FakeDatetime(2023, 6, 5, 19, 0),
    'override': dict({
      'action': 'NOT_ACTIVE',
    }),
    'restricted_reason': 'WEEK_SCHEDULE',
  })
# ---
# name: test_mower_snapshot[positions]
  list([
  ])
# ---
# name: test_mower_snapshot[settings]
  dict({
    'cutting_height': None,
    'headlight': dict({
      'mode': None,
    }),
  })
# ---
# name: test_mower_snapshot[statistics]
  dict({
    'cutting_blade_usage_time': None,
    'number_of_charging_cycles': None,
    'number_of_collisions': None,
    'total_charging_time': None,
    'total_cutting_time': None,
    'total_drive_distance': None,
    'total_running_time': None,
    'total_searching_time': None,
  })
# ---
# name: test_mower_snapshot[stay_out_zones]
  None
# ---
# name: test_mower_snapshot[system]
  dict({
    'model': 'Automower Aspire',
    'name': 'Test Mower 1',
    'serial_number': 123,
  })
# ---
# name: test_mower_snapshot[work_areas]
  None
# ---
